include/oneapi/dpl/experimental/kt/two_pass_scan.h

```diff
@@ -319,45 +418,61 @@ two_pass_scan(sycl::queue q, _InRng&& __in_rng, _OutRng&& __out_rng,
-                    for (int i = 0; i < (g >> log2_VL) + 1; i++)
+                    if (g > 0)
                     {
-                        v = tmp_storage[i * num_sub_groups_local * VL +
-                                        (num_sub_groups_local * sub_group_local_id + offset)];
-                        std::tie(v, carry) = sub_group_scan<VL, true>(sub_group, std::move(v), binary_op, carry);
-                        if (i != (g >> log2_VL))
-                            carry_last = carry;
+                        // only need the last element from each scan of num_sub_groups_local subgroup reductions
+                        const auto elements_to_process = subgroups_before_my_group / num_sub_groups_local;
+                        const auto pre_carry_iters = oneapi::dpl::__internal::__dpl_ceiling_div(elements_to_process, VL);
+                        if (pre_carry_iters == 1)
+                        {
+                            // single partial scan
+                            auto proposed_idx = num_sub_groups_local * sub_group_local_id + offset;
+                            auto remaining_elements = elements_to_process;
+                            auto reduction_idx = (proposed_idx < subgroups_before_my_group) ? proposed_idx : subgroups_before_my_group - 1;
+                            value.__setup(tmp_storage[reduction_idx]);
+                            sub_group_scan_partial<VL, true, false>(sub_group, value.__v, binary_op, carry_last, remaining_elements);
+                        }
+                        else
+                        {
+                            // multiple iterations
+                            // first 1 full
+                            value.__setup(tmp_storage[num_sub_groups_local * sub_group_local_id + offset]);
+                            sub_group_scan<VL, true, false>(sub_group, value.__v, binary_op, carry_last);
+
+                            // then some number of full iterations
+                            for (int i = 1; i < pre_carry_iters - 1; i++)
+                            {
+                                auto reduction_idx = i * num_sub_groups_local * VL + num_sub_groups_local * sub_group_local_id + offset;
+                                value.__v = tmp_storage[reduction_idx];
+                                sub_group_scan<VL, true, true>(sub_group, value.__v, binary_op, carry_last);
+                            }
+
+                            // final partial iteration
+                            auto proposed_idx = (pre_carry_iters - 1) * num_sub_groups_local * VL + num_sub_groups_local * sub_group_local_id + offset;
+                            auto remaining_elements = elements_to_process - ((pre_carry_iters - 1) * VL);
+                            auto reduction_idx = (proposed_idx < subgroups_before_my_group) ? proposed_idx : subgroups_before_my_group - 1;
+                            value.__setup(tmp_storage[reduction_idx]);
```

Comment:
Should we call `__setup` here? `value` has already been constructed here.

Reply:
good catch.  fixed.